observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -0,0 +1,436 @@
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  /**
+   *  Class for proto related helper methods.
+   */
+  static final class LogSinkWriter {
+    private final Sink sink;
+    private final TimeProvider timeProvider;
+
+    LogSinkWriter(Sink sink, TimeProvider timeProvider) {
+      this.sink = sink;
+      this.timeProvider = timeProvider;
+    }
+
+    GrpcLogRecord.Builder createTimestamp() {
+      long nanos = timeProvider.currentTimeNanos();
+      return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
+    }
+
+
+    /**
+     * Logs the request header.
+     * Binary logging equivalent of logClientHeader.
+     */
+    void logRequestHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        String authority,
+        @Nullable Duration timeout,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        // null on client side
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_SERVER,
+          "peerAddress can only be specified by server");
+
+      PayloadBuilder<GrpcLogRecord.Metadata.Builder> pair = createMetadataProto(metadata);
+      GrpcLogRecord.Builder logEntryBuilder = createTimestamp()
+          .setSequenceId(seqId)
+          .setServiceName(serviceName)
+          .setMethodName(methodName)
+          .setAuthority(authority)
+          .setEventType(EventType.GRPC_CALL_REQUEST_HEADER)
+          .setEventLogger(eventLogger)
+          .setLogLevel(LogLevel.LOG_LEVEL_DEBUG)
+          .setMetadata(pair.proto)
+          .setPayloadSize(pair.size)
+          .setRpcId(rpcId);
+      if (timeout != null) {
+        logEntryBuilder.setTimeout(timeout);
+      }
+      if (peerAddress != null) {
+        logEntryBuilder.setPeerAddress(socketAddressToProto(peerAddress));
+      }
+      logger.log(Level.FINEST, "Writing Request Header to GcpLogSink");
+      sink.write(logEntryBuilder.build());
+    }
+
+    /**
+     * Logs the reponse header.
+     * Binary logging equivalent of logServerHeader.
+     */
+    void logResponseHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_CLIENT,
```

Comment:
Similar to above: would like to see some explanation of why the peerAddress is null on the server (when sending response header). Is it tested and is this true?

Reply:
As per our discussion offline added a comment, saying we will only be logging the first instance of the peer Address being received. The translates to Response header event on client side and Request header event on server side. And for the trailer it will be logged on client side only to keep it as same as Binary Logging behavior for now.